target: use mrw generated attribute header files

devenrao opened the pull request:
Switch to using the generated attribute-related header files produced during the build process, replacing previously hand-coded versions. This ensures consistency with the build system’s definitions and reduces manual maintenance.


Change-Id: Ie74a64bee6054f2f7e5fa40419bee6f6878c5bf5

devenrao committed:
target: use mrw generated attribute header files

Switch to using the generated attribute-related header files produced
during the build process, replacing previously hand-coded versions.
This ensures consistency with the build system’s definitions and
reduces manual maintenance.

pdbg_refactor$ meson test -C build
ninja: no work to do.
1/1 targeting-tests        OK              0.01s
Ok:                 1
Expected Fail:      0
Fail:               0
Unexpected Pass:    0
Skipped:            0
Timeout:            0

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: Ie74a64bee6054f2f7e5fa40419bee6f6878c5bf5

devenrao committed:
target: add compile-time checks for attribute access

Enforce compile-time validation of attribute access in Target::getAttr
and Target::setAttr based on the readable and writeable flags defined in
AttributeTraits. This ensures that:

- Attributes marked as readable can only be accessed via get methods.
- Attributes marked as writeable can only be modified via set methods.

Also added unit tests to verify that invalid get/set usage produces
compile-time errors.

[----------] 18 tests from TargetServiceTest
[ RUN      ] TargetServiceTest.TestFDTNotNull
[ RUN      ] TargetServiceTest.CanReadReadableAttr
[       OK ] TargetServiceTest.CanReadReadableAttr (0 ms)
[ RUN      ] TargetServiceTest.CanWriteWritableAttr
[       OK ] TargetServiceTest.CanWriteWritableAttr (0 ms)
[----------] 18 tests from TargetServiceTest (1 ms total)
Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I1fe555091c2b630ebc641432aed6f56c9d5cf36f

devenrao committed:
target: add Doxygen comments for target classes

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I2d419da67725b08388893f7cfb697e31c01143f0

gcpin commented on target/targeting/xmltohb/attributetraits.H:
```diff
@@ -323,37 +477,2378 @@ template <> class AttributeTraits<ATTR_HWACCESS_METHOD>
+        notFspMutex,
+        fspAccessible
+    };
+    typedef uint32_t Type;
```
Is it safe to assume that pointers are of 32-bit size? If the memory addresses are 64-bit long, we may see loss of data. Should we use uintptr_t instead to be sure that it is not system specific.

devenrao replied:
this is hostboot parsing logic more of question to deepa , will ask her

devenrao committed:
target: separate targeting and service libraries

Split the targeting and target service code into separate libraries
to avoid circular dependencies when adding the hardware access layer.

pdbg_refactor$ meson test -C build
ninja: Entering directory `/home/xxx/zopenbmc_work/target0609/pdbg_refactor/build'
ninja: no work to do.
1/1 targeting-tests        OK              0.01s

Ok:                 1
Expected Fail:      0
Fail:               0
Unexpected Pass:    0
Skipped:            0
Timeout:            0

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: If913a7193bca470b6c3f268f870c0dfb9f71092a

devenrao committed:
target:update mrw generated attribute header files

Switch to using updated generated attribute-related header files
produced during the build process, replacing previously hand-coded
versions.

This ensures consistency with the build system’s definitions and
reduces manual maintenance.

pdbg_refactor$ meson test -C build
ninja: no work to do.
1/1 targeting-tests        OK              0.01s
Ok:                 1
Expected Fail:      0
Fail:               0
Unexpected Pass:    0
Skipped:            0
Timeout:            0

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I002f74f144b29b588dbc3be3da102538a7eb8d88

devenrao committed:
target: add compile-time handling for volatile attributes

- All volatile attributes now marked with `isvolatile` enum in
  AttributeTraits.
- Updated get/set attribute methods to use compile-time checks
  for volatile attributes instead of runtime `isVolatileAttr()`.
- Volatile attributes must be explicitly set before access;
  otherwise a runtime exception is thrown.
- Added GTest unit tests to validate get/set behavior for
  volatile attributes (including unset, set, and overwrite cases).

Tested:
added additional test cases
[ RUN      ] TargetServiceTest.GetVolatileBeforeSetThrows
[       OK ] TargetServiceTest.GetVolatileBeforeSetThrows (0 ms)
[ RUN      ] TargetServiceTest.SetAndGetVolatileWorks
[       OK ] TargetServiceTest.SetAndGetVolatileWorks (0 ms)
[ RUN      ] TargetServiceTest.getAttrPrefersVolatileOverFdt
[       OK ] TargetServiceTest.getAttrPrefersVolatileOverFdt (0 ms)

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: If080845be62c065a07c03495e76305ac00397e7a

gcpin commented on target/service/meson.build:
```diff
@@ -0,0 +1,20 @@
+    install: true,
+)
+
+# Export public headers into /usr/include/target
```
they'll be installed to /usr/include/targeting right?

gcpin commented on target/targeting/meson.build:
```diff
@@ -0,0 +1,33 @@
+    install: true,
+)
+
+# Export public headers into /usr/include/target
```
same here ... they'll be installed to /usr/include/targeting right?

gcpin commented on target/targeting/target_tmpl.H:
```diff
@@ -148,14 +147,56 @@ bool Target::trySetAttr(
+template <const ATTRIBUTE_ID A>
+void Target::setVolatileAttr(const typename AttributeTraits<A>::Type& attrValue)
+{
+    (void)AttributeTraits<A>::writeable;
```
should a volatile attribute be writeable to device tree? if an attribute is volatile it could be both readable & writable to locally cached copy irrespective of the original attribute being readable/writeable right ?

when setVolatileAttr is called, we dont intend to write to device tree , isnt it?

gcpin commented on target/targeting/target_tmpl.H:
```diff
@@ -148,14 +147,56 @@ bool Target::trySetAttr(
+typename AttributeTraits<A>::Type Target::getVolatileAttr() const
+{
+    // compile-time validation
+    (void)AttributeTraits<A>::readable;
```
same as below comment, volatile is self sufficient right ? need not be combined with readable/writeable isnt it?

devenrao committed:
target: rename service folder to targetsvc

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I2fb5b88f6f5cf3484ffd827d5cd7c2c6166558e2
